CRUD operations for jobDetails

jinsim95 opened the pull request.

jinsim95 commented:
@mattfan00 @jonako99 Hey guys, so I am working on my task "CRUD operations of job deatils" and I am stuck on something and I was hoping to get some help/tips from you guys. So I've made the basic structure for the CRUD operations in 'app.js' but am having trouble how to retrieve a particular job's ObjectID. Thanks to Matt's examples, I am trying to use the ":id" tag in the URL's to retrieve one particular job from the database and either modify or delete it using app.put and app.delete. My goal is to access a particular job from the business dashboard, click on "View Applications," and either delete the job or adjust its status using the drop-down menu within the "View Applications" page I am also having a little trouble understanding how the rendering of this view applications page works so I've tagged Jona as well. I'll continue to look into this and try and find how to retrieve and use the current job's ID but I'd really appreciate any sort of help or suggestions.

mattfan00 commented:
Hey Jin, the backend routes look good!

So the first thing you will have to do is render all of the jobs on the dashboard page. It seems there is a route that you made that already retrieves all of the jobs according to the company and returns it as an array. So this array will look something like: 

```
[
  {
     "_id": some random string,
     "company": "amazon",
     "jobTitle": "software engineer",
     ...
  },
  ...
}
```

On the dashboard, you will want to loop through each of the entries in this array and render the appropriate information per job in the list. The key thing is the link to the "View Applications". Right now the link is being rendered the same for all of the jobs on the frontend. This should be changed to a url such as `/dashboard/:id`, where `:id` is the id of the job you are currently rendering. This way, when you click on the "View Applications" link, the webpage of `http://localhost:3000/dashboard/:id` will show up. As mentioned before, this `:id` field will be the unique identifier of the job you are trying to retrieve. So now, you can add a `useEffect` hook that will send a request to `http://localhost:4000/business/jobDetails/:id` (or whatever the route is to get the details of a specific job). 

So there will probably need to be a couple of changes made on the front end in order to make this happen. Currently the route to the "View Applications" link is `/dashboard/applications`. This will probably need to be changed to `/dashboard/:id` or maybe `/dashboard:/id/applications` (prob doesn't matter too much). Additionally, right now there are props being passed along when clicking on the "View Applications" link, this is how the "View Applications" page knows what information to render. The new way of getting the information should be through the `useEffect` hook that gets it based on the id, mentioned before. 

@ando-huang is going to be implementing something very similar to this, except for viewing the information of a specific applicant!

jinsim95 commented:
@mattfan00 Thank you so much for the detailed comments! I've made some changes so that clicking 'View Applications' for one of the jobs would take the user to the link associated with that particular job, i.e. '/business/dashboard/:id_of_the_job." 
@jonako99 I don't have a very good understanding of the front-end code and am still having trouble understanding where exactly in the code the rendering of the 'View Applications' page is happening. I feel like I need to replace 'props.job' with the job object retrieved from the database (by sending a GET request to the link: '/jobs/:id_of_the_job') but I just can't seem to exactly understand where and how the rendering of the page is happening. Sorry to bother you with this Jona, but could you clarify this for me?

jinsim95 commented:
Thanks guys for all the help. I think I've got it now, although sometimes changing the status of the job acts a bit laggy and needs refreshing the page. Otherwise, it should be fine, I think.

mattfan00 commented on back-end/app.js:
```diff
@@ -61,14 +60,41 @@ app.get("/", (req, res) => {
+    {status: req.body.changeStatus},
+    {returnOriginal: false}
+  )
+})
```
I believe the reason why there is a slight delay with the update command is because there isn't a `res.json` command here. So whenever the frontend sends over a request to update the job, it will expect a response back (once it receives the response back, it will then update the text on the frontend), but since there is no `res.json`, a response is never sent back and the frontend is stuck waiting for the response. So you can put something like `res.json("Successfully updated")` or maybe just send back the updated job, it probably shouldn't matter too much! :)

mattfan00 reviewed (commented):
I added one comment about why I think there is a delay with the update command! Additionally (and I think this is my bad because I accidentally put a route in the `app.js` file), do you mind moving the routes that you created, into one of the files in the `routes` directory. Maybe `routes/getBusinessJobs.js`? 

Everything works great thoughhhh :)

jinsim95 commented:
Hey Matt, thanks for all the help I kinda feel bad for always asking you so many questions...(;
(I think) I have fixed all the issues previously mentioned, including laggy status change for a job and I also realized that Create New Job wasn't actually creating and saving a new job to the DB so I've implemented that part as well. 
I've tested for a few times and checked that the jobs were getting added and that they were getting displayed on the dashboard, so it should be okay I think.

mattfan00 commented:
Hahahaha np about the questions :) Yahhh this looks good, nice!!!!
